### PR TITLE
EIP86 - CREATE2

### PR DESCRIPTION
**Pull request by yann300**

### CODE REVIEW

File: libethereum/Executive.cpp [hunk 1]
```diff
@@ -306,9 +306,30 @@ bool Executive::call(CallParameters const& _p, u256 const& _gasPrice, Address co
 	return !m_ext;
 }
 
-bool Executive::create(Address _sender, u256 _endowment, u256 _gasPrice, u256 _gas, bytesConstRef _init, Address _origin)
+bool Executive::create(Address _txSender, u256 _endowment, u256 _gasPrice, u256 _gas, bytesConstRef _init, Address _origin)
```
Comment (pirapira) [inline libethereum/Executive.cpp]: Maybe use `u256 const&` because now the arguments are passed deeper and deeper.  Also `Address const&`.

**Comment by gumb0** [comment]
Please rebase to fix linux build on travis

**Comment by winsvega** [comment]
should work now. 
we could merge this. and then add more tests for CREATE2, ZeroTransaction and CreateTransactions

File: libethereum/Executive.cpp [hunk 2]
```diff
@@ -363,9 +363,9 @@ bool Executive::executeCreate(Address _sender, u256 _endowment, u256 _gasPrice,
 	}
 	else
 	{
-		if ((m_s.addressHasCode(m_newAddress) || m_s.getNonce(m_newAddress) != 0))
+		if ((m_s.addressHasCode(m_newAddress) || m_s.getNonce(m_newAddress) > 1))
```
Comment (gumb0) [inline libethereum/Executive.cpp]: @winsvega  I don't get this change, this means we won't fail when nonce == 1 ?

File: libethereum/Executive.cpp [hunk 3]
```diff
@@ -331,12 +352,25 @@ bool Executive::create(Address _sender, u256 _endowment, u256 _gasPrice, u256 _g
...
+	{
+		if (m_s.addressHasCode(m_newAddress))
+			// Overwrite with empty code in case the account already has a code
+			// (address collision -- not real life case but we can check it with
+			// synthetic tests).
+			m_s.setCode(m_newAddress, {});
+	}
+	else
+	{
+		if ((m_s.addressHasCode(m_newAddress) || m_s.getNonce(m_newAddress) > 1))
```
Comment (gumb0) [inline libethereum/Executive.cpp]: @winsvega I don't get this change, this means we won't fail when nonce == 1 ?

File: test/tools/jsontests/vm.cpp [hunk 4]
```diff
@@ -38,9 +39,28 @@ FakeExtVM::FakeExtVM(EnvInfo const& _envInfo, unsigned _depth):			/// TODO: XXX:
 	ExtVMFace(_envInfo, Address(), Address(), Address(), 0, 1, bytesConstRef(), bytes(), EmptySHA3, false, _depth)
 {}
 
-std::pair<h160, eth::owning_bytes_ref> FakeExtVM::create(u256 _endowment, u256& io_gas, bytesConstRef _init, OnOpFunc const&)
+std::pair<h160, eth::owning_bytes_ref> FakeExtVM::create(u256 _endowment, u256& io_gas, bytesConstRef _init, Instruction _op, u256 salt, OnOpFunc const&)
 {
-	Address na = right160(sha3(rlpList(myAddress, get<1>(addresses[myAddress]))));
+	unique_ptr<SealEngineFace> se(ChainParams(genesisInfo(eth::Network::MainNetworkTest)).createSealEngine());
```
Comment (gumb0) [inline test/tools/jsontests/vm.cpp]: This change should be removed, nobody needs it, `FakeExtVM` is used only with frontier rules

File: libethereum/Executive.cpp (hunk 3 above)
Comment (winsvega) [inline libethereum/Executive.cpp]: hmm. thats strange.

Comment (winsvega) [inline libethereum/Executive.cpp]: I see so here is the issue. the new address is being created with default nonce = 1. 
then it is being checked for existance. this is not a reliable way to check that the address is already exist.

Comment (pirapira) [inline libethereum/Executive.cpp]: The EIP [says](https://github.com/ethereum/EIPs/pull/208/files#diff-95b306e6e6d2d6dbb0c5fd3ac4ea45c4R28), check for `hasCode || nonce != 0`.

(OK, it says "non-empty nonce" which is weird.)

Comment (winsvega) [inline libethereum/Executive.cpp]: the EIP is correct. however at that point in code the adress is somehow already present at the state
as if we create an address then check if its existing

Comment (gumb0) [inline libethereum/Executive.cpp]: Ah I see, we've already incremented the nonce on line 350

Comment (gumb0) [inline libethereum/Executive.cpp]: Then I guess this check should better be done before nonce incrementation. Probably just after incrementing sender's nonce

File: libethereum/Executive.cpp [hunk 5]
```diff
@@ -331,12 +353,25 @@ bool Executive::create(Address _sender, u256 _endowment, u256 _gasPrice, u256 _g
...
+			// synthetic tests).
+			m_s.setCode(m_newAddress, {});
+	}
+	else
+	{
+		if (accountAlreadyExist)
+		{
+			clog(StateSafeExceptions) << "Address already used: " << m_newAddress;
+			m_gas = 0;
+			m_excepted = TransactionException::AddressAlreadyUsed;
```
Comment (pirapira) [inline libethereum/Executive.cpp]: Should we also clear `m_newAddress`?

Comment (winsvega) [inline libethereum/Executive.cpp]: I guess revert do the logic

Comment (pirapira) [inline libethereum/Executive.cpp]: Yes it does.

File: libethereum/Executive.cpp [hunk 6]
```diff
@@ -331,12 +353,25 @@ bool Executive::create(Address _sender, u256 _endowment, u256 _gasPrice, u256 _g
...
+			revert();
```
Comment (pirapira) [inline libethereum/Executive.cpp]: In this case, I think true must be returned from this function, indicating that the VM execution is not necessary.

File: libethereum/Executive.h [hunk 7]
```diff
@@ -150,10 +149,12 @@ class Executive
 
 	/// Set up the executive for evaluating a bare CREATE (contract-creation) operation.
 	/// @returns false iff go() must be called (and thus a VM execution in required).
-	bool create(Address _txSender, u256 _endowment, u256 _gasPrice, u256 _gas, bytesConstRef _code, Address _originAddress);
+	bool create(Address const& _txSender, u256 const& _endowment, u256 const& _gasPrice, u256 const& _gas, bytesConstRef _code, Address const& _originAddress);
+	bool createOpcode(Address const& _sender, u256 const& _endowment, u256 const& _gasPrice, u256 const& _gas, bytesConstRef _code, Address const& _originAddress);
+	bool create2Opcode(Address const& _sender, u256 const& _endowment, u256 const& _gasPrice, u256 const& _gas, bytesConstRef _code, Address const& _originAddress, u256 const& _salt);
```
Comment (pirapira) [inline libethereum/Executive.h]: Please document the return value of the two new functions (it is not obvious).

File: libethereum/Executive.h [hunk 8]
```diff
@@ -183,6 +184,8 @@ class Executive
 	void revert();
 
 private:
+	bool executeCreate(Address const& _txSender, u256 const& _endowment, u256 const& _gasPrice, u256 const& _gas, bytesConstRef _code, Address const& _originAddress);
```
Comment (pirapira) [inline libethereum/Executive.h]: Please document the return value.

File: libethereum/Executive.cpp (hunk 6 above)
Comment (winsvega) [inline libethereum/Executive.cpp]: ask @yann300  about this

Comment (pirapira) [inline libethereum/Executive.cpp]: I made the change.

**Comment by pirapira** [comment]
@winsvega please review my commit.

**Pull request merged**